### PR TITLE
GH-118093: Specialize `CALL_KW`

### PR DESCRIPTION
**Pull request by markshannon**
According to [our stats](https://github.com/faster-cpython/benchmarking/blob/main/results/bm-20240810-3.14.0a0-363374c-PYTHON_UOPS/bm-20240810-azure-x86_64-python-363374cf69a7e2292fe3-3.14.0a0-363374c-pystats.md) the `CALL_KW` opcode is one of the most common opcodes that prevents projection of tier 2 traces.
Making `CALL_KW` suitable for tier 2 is infeasible, but we can specialize it in tier 1 such that the specializations can be handled by tier 2.

This PR adds three specializations:
* `CALL_KW_PY` for calls to Python functions
* `CALL_KW_BOUND_METHODS` for calls to bound methods
* `CALL_KW_NON_PY` for calls to everything else.

* Issue: gh-118093

### CODE REVIEW

**Comment by markshannon** [comment]
No impact on [performance](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240814-3.14.0a0-6b9055c-PYTHON_UOPS)
[Stats] show 99.9% specialization with a low (<1%) miss rate.

File: Lib/test/test_dis.py [hunk 1]
```diff
@@ -257,6 +257,9 @@ def wrap_func_w_kwargs():
               LOAD_CONST               3 (5)
               LOAD_CONST               4 (('c',))
               CALL_KW                  3
+              CACHE
```
Comment (Fidget-Spinner) [inline Lib/test/test_dis.py]: AFAIK, CACHE is not supposed to show up. Just like LOAD_GLOBAL has no CACHE showing up after. We need to teach dis about CALL_KW. There used to be a CACHE entry list in `opcode.py` but I don't know if we still maintain it there. Maybe @iritkatriel knows?

File: Python/bytecodes.c [hunk 2]
```diff
@@ -4104,8 +4112,144 @@ dummy_func(
...
+            PyObject *callable_o = PyStackRef_AsPyObjectBorrow(callable);
+            PyObject *self_or_null_o = PyStackRef_AsPyObjectBorrow(self_or_null);
+
+            int total_args = oparg;
+            if (self_or_null_o != NULL) {
+                args--;
+                total_args++;
+            }
+            /* Callable is not a normal Python function */
+            STACKREFS_TO_PYOBJECTS(args, total_args, args_o);
```
Comment (Fidget-Spinner) [inline Python/bytecodes.c:4226]: Do the kw args lie on the stack after the `args[oparg]` as well? If so, you need to convert that too. If not then just ignore this review.

**Review by Fidget-Spinner — COMMENTED** [review]
Looks good in general. Two questions.

File: Lib/test/test_dis.py (hunk 1 above)
Comment (markshannon) [reply Lib/test/test_dis.py]: Fixed. `CALL_KW` needed to be added to the cache table in opcode.py

File: Python/bytecodes.c (hunk 2 above)
Comment (markshannon) [reply Python/bytecodes.c:4226]: `kwnames` are not part of the `args` array. They are passed a separate argument.

**Review by Fidget-Spinner — APPROVED** [review]

File: Include/internal/pycore_code.h [hunk 3]
```diff
@@ -335,6 +336,8 @@ extern void _Py_Specialize_StoreSubscr(_PyStackRef container, _PyStackRef sub,
                                        _Py_CODEUNIT *instr);
 extern void _Py_Specialize_Call(_PyStackRef callable, _Py_CODEUNIT *instr,
                                 int nargs);
+extern void _Py_Specialize_CallKw(_PyStackRef callable, _Py_CODEUNIT *instr,
+                                int nargs);
```
Comment (iritkatriel) [inline Include/internal/pycore_code.h]:
```suggestion
extern void _Py_Specialize_CallKw(_PyStackRef callable, _Py_CODEUNIT *instr,
                                  int nargs);
```

**Review by iritkatriel — APPROVED** [review]

**Pull request merged**